### PR TITLE
Refactor branch state to be faster

### PR DESCRIPTION
**Pull request by jonathanwgoodwin**
Improved Query speed
Added Repo Builds for each module build because we want build trigger information.
@andyhuang91 Still a bit of a WIP, but this brought the query time from 100+s to ~500ms for Email.

This PR also adds the associated repo build (carried in the BuildInfo Object) with a given moduleBuild which will allow the UI to know what the dep tree for the build is as well as the information about the build Trigger.

### CODE REVIEW

File: BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java [hunk 1]
```diff
@@ -44,21 +42,45 @@
...
+      "ORDER BY lastSuccessfulBuild.id DESC LIMIT 1")
+  Optional<BuildInfo> getLastSuccessfulBuildInfo(@Bind("moduleId") int moduleId);
+
+  @SingleValueResult
+  @SqlQuery("" +
+      "SELECT repositoryBuild.*, lastNonSkippedBuild.* " +
+      "FROM module_builds AS lastNonSkippedBuild " +
+      "LEFT OUTER JOIN repo_builds AS repositoryBuild on (lastNonSkippedBuild.repoBuildId = repositoryBuild.id) " +
+      "WHERE lastNonSkippedBuild.moduleId = :moduleId " +
+      "AND lastNonSkippedBuild.state in ('QUEUED', 'WAITING_FOR_UPSTREAM_BUILD', 'LAUNCHING', 'IN_PROGRESS', 'SUCCEEDED', 'CANCELLED', 'FAILED') " +
```
Comment (jonathanwgoodwin) [inline BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java]: Previously it was a `!= 'skipped'` which is _much_ slower than this.
This does mean that if we add a state we need to update this list.

Comment (jonathanwgoodwin) [inline BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java]: Potential tweak here would be to have a subquery here to get the possible states and pull out skipped so we don't have to update this query in future.

File: BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java [hunk 2]
```diff
@@ -36,13 +52,40 @@ public StateService(StateDao stateDao, ModuleService moduleService) {
...
+        LOG.error("Getting module state was for branch {} encountered an exception", branchId, e);
       }
     }
+    metricRegistry.timer(METRIC_FETCH_BRANCH_STATE_TIMER).update(System.currentTimeMillis() - start, TimeUnit.MILLISECONDS);
+    LOG.debug("Got composite state in {}", System.currentTimeMillis() - start);
     return states;
   }
+
+  private CompletableFuture<ModuleState> getModuleState(Module module) {
+    return CompletableFuture.supplyAsync(() -> {
```
Comment (jonathanwgoodwin) [inline BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java]: Not sure If I should be supplying my own `Executor` here this uses `ForkJoinPool#commonPool()` by default.

Comment (jhaber) [inline BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java]: I think @axiak has a document somewhere about this, but the tl;dr is you don't want anything making network calls like this to be on the common `ForkJoinPool`

Comment (jonathanwgoodwin) [inline BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java]: Yeah pranav Mentioned to me today there was some stuff about this in our javaguides so I'll have a read there too.

File: BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java [hunk 3]
```diff
@@ -44,21 +42,36 @@
...
-      "  ORDER BY lastNonSkippedBuild.id DESC, lastSuccessfulBuild.id DESC LIMIT 1")
-  Optional<ModuleState> getModuleStateById(@BindWithRosetta Module module);
+      // last build
+      "     LEFT OUTER JOIN module_builds AS lastModuleBuild ON (module.lastBuildId = lastModuleBuild.id) " +
+      "     LEFT OUTER JOIN repo_builds AS lastRepoBuild ON (lastModuleBuild.repoBuildId = lastRepoBuild.id) " +
+      // in progress build
+      "     LEFT OUTER JOIN module_builds AS inProgressModuleBuild ON (module.inProgressBuildId = inProgressModuleBuild.id) " +
+      "     LEFT OUTER JOIN repo_builds AS inProgressRepoBuild ON (inProgressModuleBuild.repoBuildId = inProgressRepoBuild.id) " +
+      // pending build
+      "     LEFT OUTER JOIN module_builds AS pendingModuleBuild ON (module.pendingBuildId =pendingModuleBuild.id) " +
```
Comment (andyhuang91) [inline BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java]: white space around `=pendingModuleBuild.id`

File: BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java [hunk 4]
```diff
@@ -44,21 +42,36 @@
...
+      "     LEFT OUTER JOIN repo_builds AS lastNonSkippedRepoBuild ON (lastNonSkippedModuleBuild.repoBuildId = lastNonSkippedRepoBuild.id) " +
+      // last successful
+      "     LEFT OUTER JOIN module_builds AS lastSuccessfulModuleBuild ON (module.id = lastSuccessfulModuleBuild.moduleId) " +
+      "     LEFT OUTER JOIN repo_builds AS lastSuccessfulRepoBuild ON (lastSuccessfulModuleBuild.repoBuildId = lastSuccessfulRepoBuild.id) " +
+      "  WHERE module.id = :moduleId " +
+      // lastNonSkipped and lastSuccessful ids are gotten in subqueries to avoid sorting for the lastNonSkipped and then also lastSuccessful
+      "  AND lastNonSkippedModuleBuild.id = (SELECT lastNonSkippedBuild.id FROM module_builds AS lastNonSkippedBuild " +
+      "                                      LEFT OUTER JOIN repo_builds AS repositoryBuild ON (lastNonSkippedBuild.repoBuildId = repositoryBuild.id) " +
+      "                                      WHERE lastNonSkippedBuild.moduleId = :moduleId " +
+      "                                      AND lastNonSkippedBuild.state IN ('QUEUED', 'WAITING_FOR_UPSTREAM_BUILD', 'LAUNCHING', 'IN_PROGRESS', 'SUCCEEDED', 'CANCELLED', 'FAILED') " +
```
Comment (andyhuang91) [inline BlazarData/src/main/java/com/hubspot/blazar/data/dao/StateDao.java]: I'm thinking that we don't want to show 'QUEUED' (and possibly not 'WAITING_FOR_UPSTREAM_BUILD') in as the "current" build in the new UI?

File: BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java [hunk 5]
```diff
@@ -14,13 +18,18 @@
...
   private final StateDao stateDao;
   private ModuleService moduleService;
+  private MetricRegistry metricRegistry;
 
   @Inject
-  public StateService(StateDao stateDao, ModuleService moduleService) {
+  public StateService(StateDao stateDao, ModuleService moduleService, MetricRegistry metricRegistry) {
     this.stateDao = stateDao;
     this.moduleService = moduleService;
+    this.metricRegistry = metricRegistry;
```
Comment (andyhuang91) [inline BlazarData/src/main/java/com/hubspot/blazar/data/service/StateService.java]: Is this actually used anywhere?

**Pull request merged**